sound/soc/codecs/rt1318-sdw.c

```diff
@@ -0,0 +1,955 @@
+	prop->sink_dpn_prop = devm_kcalloc(&slave->dev, nval,
+		sizeof(*prop->sink_dpn_prop), GFP_KERNEL);
+	if (!prop->sink_dpn_prop)
+		return -ENOMEM;
+
+	j = 0;
+	dpn = prop->sink_dpn_prop;
+	addr = prop->sink_ports;
+	for_each_set_bit(bit, &addr, 32) {
+		dpn[j].num = bit;
+		dpn[j].type = SDW_DPN_FULL;
+		dpn[j].simple_ch_prep_sm = true;
+		dpn[j].ch_prep_timeout = 10;
+		j++;
+	}
+
+	/* set the timeout values */
+	prop->clk_stop_timeout = 20;
+
+	return 0;
+}
+
+static int rt1318_spk_coeff_init(struct rt1318_sdw_priv *rt1318)
+{
+	int ret = 0, set_flag;
+	char filename[128];
+	const char *vendor, *product;
+	struct device *dev = &rt1318->sdw_slave->dev;
+	const struct firmware *fw = NULL;
+	char *param_ptr;
+	unsigned int i, reg, data;
+
+	ret = regmap_read(rt1318->regmap, 0x3000, &set_flag);
+	if (!set_flag) {
+
+		if (rt1318->param_name == NULL) {
+			vendor = dmi_get_system_info(DMI_SYS_VENDOR);
+			product = dmi_get_system_info(DMI_PRODUCT_SKU);
+			dev_dbg(dev, "vendor:%s, product:%s\n", vendor, product);
+
+			if (vendor && product)
+				snprintf(filename, sizeof(filename),
+					"rt1318_param_%s_%s.bin", vendor, product);
+			else
+				sprintf(filename, "rt1318_param.bin");
+		} else
+			snprintf(filename, sizeof(filename), "%s", rt1318->param_name);
+
+		ret = request_firmware(&fw, filename, dev);
+		if (ret) {
+			ret = request_firmware(&fw, "rt1318_param.bin", dev);
+			if (ret) {
+				dev_warn(dev, "No binary file to load the parameters\n");
+				goto _exit_;
+			}
+		}
+
+		dev_dbg(dev, "rt1318: param fw size %zd", fw->size);
+		param_ptr = (char *)fw->data;
+		for (i = 0; i < fw->size; i += 3) {
+			reg = param_ptr[i] | (param_ptr[i + 1] << 8);
+			data = param_ptr[i + 2];
+			regmap_write(rt1318->regmap, reg, data);
+		}
+
+		regmap_write(rt1318->regmap, 0x3000, 0x01);
+		release_firmware(fw);
+	}
+
+_exit_:
+	return ret;
+}
+
+static int rt1318_io_init(struct device *dev, struct sdw_slave *slave)
+{
+	struct rt1318_sdw_priv *rt1318 = dev_get_drvdata(dev);
+
+	if (rt1318->hw_init)
+		return 0;
+
+	regcache_cache_only(rt1318->regmap, false);
+	if (rt1318->first_hw_init) {
+		regcache_cache_bypass(rt1318->regmap, true);
+	} else {
+		/*
+		 * PM runtime is only enabled when a Slave reports as Attached
+		 */
+
+		/* set autosuspend parameters */
+		pm_runtime_set_autosuspend_delay(&slave->dev, 3000);
+		pm_runtime_use_autosuspend(&slave->dev);
+
+		/* update count of parent 'active' children */
+		pm_runtime_set_active(&slave->dev);
+
+		/* make sure the device does not suspend immediately */
+		pm_runtime_mark_last_busy(&slave->dev);
+
+		pm_runtime_enable(&slave->dev);
+	}
+
+	pm_runtime_get_noresume(&slave->dev);
+
+	/* blind write */
+	regmap_multi_reg_write(rt1318->regmap, rt1318_blind_write,
+		ARRAY_SIZE(rt1318_blind_write));
+
+	if (rt1318->first_hw_init) {
+		regcache_cache_bypass(rt1318->regmap, false);
+		regcache_mark_dirty(rt1318->regmap);
+	}
+
+	if (rt1318->unattached_init)
+		regcache_sync(rt1318->regmap);
```

Comment:
use the existing flows for now, it's not that urgent to fix the unattached init stuff.

Reply:
Ok, will update.